App navigation refactor

rtibbles opened the pull request:
## Summary
* Consolidates all app bar population to follow the same pattern as bottom app bar and side nav population
* Updates the API for nav item registration to validate sub routes to allow populating the app bar
* Consolidates the components for navigation bar display and overflow behaviour
* Gets rid of navigation mixin in favour of limited composable
* Folds navComponents logic into the composable
* Gets rid of unused priority property
* Cleans up all nav item registration to remove vestiges of when they were components rather than bare objects
* Clean up Bottom App Bar navigation to ensure that something is always selected for the bottom app bar

## References
Fixes #11731 
Fixes #10395 

## Reviewer guidance
Test navigation in each of the SPAs, and also in app mode on a touch device to ensure that the bottom app bar is properly populating

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [x] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md


<!--
ELLIPSIS_HIDDEN
-->


----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 476236086f8ec0757d58208026fa637e76593871  | 
|--------|

### Summary:
Refactors app navigation by updating the registration API, introducing a `useNav` composable, and removing outdated patterns.

**Key points**:
- Refactored app navigation to use a unified pattern across the app.
- Updated API for navigation item registration to include sub-route validation.
- Replaced navigation mixins with a new `useNav` composable for better state management.
- Removed old navigation-related mixins and utilities.
- Ensured navigation components are correctly populated and responsive to changes.


----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)



<!--
ELLIPSIS_HIDDEN
-->

rtibbles committed:
Update nav component validation to validate sub routes.

rtibbles committed:
Update all sub routes to include icons.

rtibbles committed:
Fix typo and change to created from mounted.

rtibbles committed:
Update AppBar to autopopulate navigation based on registered nav items.

rtibbles committed:
Update sidenav to use useUser.
Delete unneeded logout side nav entry.

rtibbles committed:
Ensure that the bottom app navigation bar always has a defined navigation item.
Handle when it has no defined subroutes.

rtibbles committed:
Default title to empty string to prevent validation errors.

marcellamaki commented on kolibri/core/assets/src/views/CoreMenu/CoreMenuOption.vue:
```diff
@@ -52,7 +52,7 @@
             v-if="linkActive"
             v-slot="{ href, navigate, isActive }"
-            :to="{ name: subRoute.name, params: $route.params, query: $router.query }"
+            :to="{ name: subRoute.name, params: $route.params, query: $route.query }"
```
Was this just a bug we didn't catch or did something meaningfully change here?

rtibbles replied:
The former.

rtibbles committed:
Remove navComponent mixin in favour of useNav composable.

rtibbles committed:
Update all SPAs to use app bar autopopulation.

rtibbles committed:
Consolidate nav bar code.
Remove unneeded core API exports.

rtibbles committed:
Update app bar to toggle app nav behaviour via prop.
Update for useUser composable.

rtibbles committed:
Move navComponents into useNav composable.
Update apiSpec.
Make icon and url required.
Add url sort for consistent ordering.

rtibbles committed:
Clean up unused api spec entries.

marcellamaki commented on kolibri/core/assets/src/composables/useNav.js:
```diff
@@ -0,0 +1,87 @@
+    if (validateNavItem(component)) {
+      navComponents.push(component);
+    } else {
+      logging.error('Component has invalid priority, role, section, or routes');
```
Small, but I think this should be updated with the new validations checks since we are not using "priority" anymore, and maybe want to add in url or icon?

rtibbles replied:
Yes, it might be useful for error messages to actually be correct.

marcellamaki commented on kolibri/core/assets/src/composables/useNav.js:
```diff
@@ -0,0 +1,87 @@
+}
+
+function validateRoutes(routes) {
+  // Required, must be an array of objects
```
So, the url is the plugin "home" URL, and these are the sub-routes. I think this is not always required. because of things like the login/out side nav entry and profile. and the spec suggests that just url and icon are required for it to be registered

rtibbles replied:
Yes - it appears the comment is out of kilter with what the validation is actually doing, where it definitely does allow these not to be defined.

marcellamaki commented on kolibri/core/assets/src/views/AppBar.vue:
```diff
@@ -99,15 +104,16 @@
-    <!-- IF making changes to the sub nav, make sure to make -->
-    <!-- corresponding changes in SideNav.vue in regards to  -->
-    <!-- Window size and app context. Changes may need to be made -->
-    <!-- in parallel in both files for non-breaking updates -->
```
yay that this convoluted thing isn't needed anymore!

rtibbles replied:
This was mostly achieved by doing some prop-drilling instead, so that the components are controlled by the same flag at the top level in the AppBarPage.

marcellamaki commented on kolibri/core/assets/src/views/AppBar.vue:
```diff
@@ -134,36 +143,64 @@
+      const { topBarHeight, navComponents } = useNav();
+      const { isLearner, isUserLoggedIn, username, full_name } = useUser();
+      const links = computed(() => {
+        const currentComponent = navComponents.find(nc => nc.url === window.location.pathname);
```
I don't see this used elsewhere in the file, so I am just wondering if it's used here, or if this syntax means it's now also available when `links` is referenced, i.e. in `<Navbar>`? I'm confused about how this works

rtibbles replied:
Lines 155 to 166 are defining the links computed prop, so `currentComponent` is only used inside the scope of the function that computes that.

Although, `currentComponent` is not really the right word now that they aren't components any more.

marcellamaki reviewed (commented):
I think I may have more comprehension questions, but so as to get this moving again after delaying, I'll start with the questions I have, and then follow up as needed once I digest the answers :)

pcenov commented:
Hi @rtibbles, here are the issues I was able to identify:

1. All the icons in the navigation bar are displayed in white instead of black:

![icons](https://github.com/learningequality/kolibri/assets/79847249/9cddd135-3f08-4477-99a3-db3e21fad43c)

2. The navigation bar disappears completely at the Profile and My downloads pages (this is especially unpleasant in the Mac app where the user can't navigate out of these pages):

https://github.com/learningequality/kolibri/assets/79847249/83a1c23e-16b0-4789-9efa-434ec245417f

3. The side bar disappears completely when I shrink the page to less than 840 pixels:

![2024-05-08_16-30-32](https://github.com/learningequality/kolibri/assets/79847249/fff48521-6ae1-49e3-b735-986b34ca6ea2)

The same is also valid for the Mac app:

https://github.com/learningequality/kolibri/assets/79847249/a2f11dc5-6317-405e-a9ef-094b52257336

4. When I change the language to any other language, the 'Sign out' button is not localized:

![signout](https://github.com/learningequality/kolibri/assets/79847249/11342792-b83d-4713-8cfe-71a6470118bf)

rtibbles commented:
These all seem like problems - although I thought I had handled the icon colour already.

One thing to note is that because I moved the location of the sign out string, this will remain unlocalized until we've done the translation pass to reinstate the translated string.

rtibbles commented:
Hey @ellipsis, give me a code review

rtibbles committed:
Clean up naming and code comments based on review feedback.

rtibbles committed:
Update UI issues reported in QA.

rtibbles commented:
@marcellamaki have addressed comments.

@pcenov I have fixed all the reported issues (except the internationalized Sign out, which will not be fixed until we redo translations, as noted).

pcenov commented:
Hi @rtibbles - I confirm that the reported issues are fixed now. While regression testing I spotted just one more - in mobile view the ellipsis button is positioned too close to the right, it's oval-shaped when clicked and the dots are white:

https://github.com/learningequality/kolibri/assets/79847249/90e9a63e-a82d-4198-b9f7-3151b62f7306

marcellamaki reviewed (commented):
This is good to go on my side, one the small final UI updated that Peter noticed is fixed. (I am also not opposed to that being follow up if it's not a super quick fix, but it seems like it might be easier to just fix rather than open an issue for it)

rtibbles committed:
Update color for dot menu. Prevent flex shrink.

rtibbles commented:
Updated:
![image](https://github.com/learningequality/kolibri/assets/1680573/f573a3f5-1daa-4789-8fd2-02622936fa45)
